igor-web/src/main/groovy/com/netflix/spinnaker/igor/scm/stash/client/StashMaster.groovy

```diff
@@ -31,42 +35,38 @@ import javax.validation.Valid
 /**
  * Wrapper class for a collection of Stash clients
  */
-class StashMaster {
+@Slf4j
+class StashMaster extends AbstractScmMaster {
     StashClient stashClient
     String baseUrl
 
-    @Bean
-    @ConditionalOnProperty('stash.base-url')
-    StashMaster stashMaster(@Valid StashProperties stashProperties) {
-        log.info "bootstrapping ${stashProperties.baseUrl}"
-        new StashMaster(
-            stashClient : stashClient(stashProperties.baseUrl, stashProperties.username, stashProperties.password), baseUrl: stashProperties.baseUrl)
+  List<String> listDirectory(String projectKey, String repositorySlug, String path, String at) {
+    try {
+      return stashClient.listDirectory(projectKey, repositorySlug, path, at).toChildFilenames()
+    } catch (RetrofitError e) {
+      if (e.getKind() == RetrofitError.Kind.NETWORK) {
+        throw new NotFoundException("Could not find the server ${baseUrl}")
+      }
+      log.error(
+        "Failed to fetch file from {}/{}/{}, reason: {}",
+        projectKey, repositorySlug, path, e.message
+      )
+      throw e
```

Comment:
should we throw the exception here? isn't it risky?

Reply:
The general behavior when calling these retrofit clients, AFAIK, is to just let exceptions bubble up to get handled upstream. The only reason we even have this `catch` block here is to convert the network error into a not-found, so for other cases we do need to either bubble up the original exception, or convert it to another exception type (which I wasn't aware of any).